### PR TITLE
fill-in-the-middle collator

### PR DESCRIPTION
**Pull request by csjackson0**
The FIM collator subclasses DataCollatorForLanguageModeling and creates a custom collator for FIM modeling on the fly.

### CODE REVIEW

**Comment by jamaliki** [comment]
Hey Corey, thanks for this! One thing I don't understand is how we are going to work with batches. I was proposing to do this on the dataset side, but here you are doing it on the collator side, which means things are already batched and padded. As you can probably see, this means the task of picking which indices to mask, etc becomes pretty difficult. What are your thoughts?

**Comment by csjackson0** [comment]
Hi Kiarash, thank you for the feedback! The collator iterates through each sequence in a batch and applies the FIM transform. This would allow to have a mixture of both left-to-right and FIM for each batch. 

Should span_start be at least a few residues away from the start of the aa_string? I believe currently the middle_span can start at position 0 in the aa_string.

File: protein_lm/utils/fim_collator.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+            end_span_token                      # <end_span>
+        ]
+
+        inputs=torch.cat(fim_transformation)
+        return inputs
+  
+    
+    def torch_call(self, examples)-> Dict[str, Any]:
+        # Override the torch_call method to make custom fim collator.
+        tokenize = BertTokenizer.from_pretrained("bert-base-uncased")
```
Comment (jamaliki) [inline protein_lm/utils/fim_collator.py:53]: We probably want the tokenizer to be an input to the DataCollator class so that we can use other tokenizers.

File: protein_lm/utils/fim_collator.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        self.fim_freq = fim_freq
+        self.tokenizer=AptTokenizer()
+        self.min_span_len=  min_span_len
+        self.max_span_len= max_span_len
+        self.mlm=mlm
+
+    def fim_transform(self, inputs):
+        max_span_len = min(self.max_span_len, inputs.shape[-1])
+        if self.min_span_len == max_span_len:
+          span_len=self.min_span_len
```
Comment (jamaliki) [inline protein_lm/utils/fim_collator.py:31]: Can you make sure you use a consistent tab size of 4 spaces, there is a mix in this file

File: protein_lm/utils/fim_collator.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        fim_transformation = [                 
+            inputs[ :span_start],             # <cls>+ Tokens before the masked span
+            middle_span_token,                # <middle_span>
+            inputs[ span_end:],               # Tokens after the masked span + <eos>
+            inputs[ span_start:span_end], #tokens in middle
+            end_span_token                      # <end_span>
+        ]
+
+        inputs=torch.cat(fim_transformation)
+        return inputs
```
Comment (jamaliki) [inline protein_lm/utils/fim_collator.py:48]: This seems right to me, can you run the tests I did for the FIM transform function I had on this too?

File: protein_lm/utils/fim_collator.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+          if random.random() > self.fim_freq:
+            #print("start",batch["input_ids"].shape)
+            fim_batch= [self.fim_transform(inputs).unsqueeze(0) for inputs in batch['input_ids']]
+            batch['input_ids']=torch.cat(fim_batch)
+            attention_mask = torch.ones_like(batch["input_ids"])
+            batch["attention_mask"] = attention_mask
+            labels=batch['input_ids'].clone()
+            if self.tokenizer.pad_token_id is not None:
+                labels[labels == self.tokenizer.pad_token_id] = -100
+            batch["labels"] = labels
```
Comment (jamaliki) [inline protein_lm/utils/fim_collator.py:79]: One thing, for both FIM and normal, I think that the labels should be shifted by one to the right since this is an auto-regressive task.

Comment (jamaliki) [reply protein_lm/utils/fim_collator.py:79]: This would happen before padding ideally

**Review by jamaliki — CHANGES_REQUESTED** [review]
Some small changes but looks good!

File: protein_lm/utils/fim_collator.py (hunk 2 above)
Comment (csjackson0) [reply protein_lm/utils/fim_collator.py]: fixed!

File: protein_lm/utils/fim_collator.py (hunk 3 above)
Comment (csjackson0) [reply protein_lm/utils/fim_collator.py:57]: I ran the tests on the FimDataCollator and all tests pass! I added them to the fim_tests.py file.

File: protein_lm/utils/fim_collator.py (hunk 1 above)
Comment (csjackson0) [reply protein_lm/utils/fim_collator.py:62]: Updated the FimDataCollator class to accept the tokenizer as input.

File: protein_lm/utils/fim_collator.py (hunk 4 above)
Comment (csjackson0) [reply protein_lm/utils/fim_collator.py]: This seems to be done in the APTLMHeadModel class when calculating the loss.

File: protein_lm/utils/fim_collator.py (hunk 1 above)
Comment (csjackson0) [reply protein_lm/utils/fim_collator.py:62]: I am using BertTokenizer for the padding function for now.

File: protein_lm/utils/fim_collator.py (hunk 3 above)
Comment (jamaliki) [reply protein_lm/utils/fim_collator.py:57]: Awesome!

**Review by jamaliki — APPROVED** [review]
LGTM!

File: protein_lm/utils/fim_collator.py (hunk 4 above)
Comment (jamaliki) [reply protein_lm/utils/fim_collator.py]: Great

**Comment by pascalnotin** [comment]
Merged the PR into main - thank you @csjackson0  & @jamaliki !
